### DISCUSSION TITLE
Elegant way of dealing with fader range not being full 0-10k?

### DISCUSSION THREAD

**Discussion by LukeThomasDawson** [tttapa/Control-Surface-Motor-Fader#15] [Ideas]
Howdy again,

Some suggested guidance would be super appreciated on how to tackle the max/min extents of the ALPS 10k faders I am using...

It appears that the lowest and highest values are not actually GND and 10k, but rather a few hundred ohms off, so that reading in the ADC doesn't give a clean 0 - 1023 but more like 19 - 1006...  Happy to care of this in software (so that the motor isn't constantly trying to drive down to 0 and heat up for instance) but was wondering what the sophisticated **_tttapa_** approach would be in to elegantly tackle the issue?

Happy to contribute if this feature is useful for others too.

Best,
Luke

**Comment by tttapa** [comment]
I don't think you can easily fix this in hardware.

In software, the obvious choice would be to just `map()` the analog input value. However, this poses an interesting problem: since you're mapping a smaller range (19-1006) to a larger one (0-1023), some target positions in the larger range become unattainable. (As a side note, the overhead of the integer division required for this mapping might not be insignificant if you're doing it at ~1 kHz.)  
For the controller, it might therefore be better to map the reference from (0-1023) to (19-1006), rather than mapping the measurement from (19-1006) to (0-1023). You'll still have to map from (19-1006) to (0-1023) before sending the potentiometer value over MIDI. Essentially, you can move all range conversions to the MIDI part of the code, without modifying the controller.

**Reply by LukeThomasDawson** [reply]
Yep, love it.

The immediate thought was to jump straight in to a `map()` on the analog read, but really what is needed here is the _setpoint_ to be the one where realigning happens so that the pot actually does get set to a real value that it can achieve - and then the MIDI setpoint + value return can do the adjustments. 

I'll add in some pot max/min variables on the MIDI side and test out some different ways of "learning" what the true max and min values are for a given fader.

Thanks for hot tip on int division too... I'll keep an eye on cycles : )

Best,
Luke
